### ISSUE TITLE
Beta: Typescript types aren't being resolved properly for imports

### ISSUE THREAD

**Issue by jgentes** [katspaugh/wavesurfer.js#2877]
I thought there was a fix for this one, but it seems the types aren't resolving properly:
![image](https://github.com/katspaugh/wavesurfer.js/assets/1149470/592c1de5-a239-4297-a771-5d5936107062)
It's giving me:
```
Could not find a declaration file for module 'wavesurfer.js'. 'd:/mixpoint/node_modules/wavesurfer.js/dist/wavesurfer.min.js' implicitly has an 'any' type.
  Try `npm i --save-dev @types/wavesurfer.js` if it exists or add a new declaration (.d.ts) file containing `declare module 'wavesurfer.js';`ts(7016)
```
This also happens with the normal WaveSurfer import:
![image](https://github.com/katspaugh/wavesurfer.js/assets/1149470/5eedbd4c-3847-4b64-a4a0-adb2518ab881)

Thanks!

**Comment by jgentes** [comment]
Here's my tsconfig.json for context:
```js
{
  "compilerOptions": {
    "lib": ["DOM", "DOM.Iterable", "ES2020"],
    "isolatedModules": true,
    "jsx": "react-jsx",
    "resolveJsonModule": true,
    "module": "NodeNext",
    "moduleResolution": "node16",
    "baseUrl": ".",
    "noEmit": true,
    "paths": {
      "~/*": ["./app/*"]
    },
    "allowJs": true,
    "forceConsistentCasingInFileNames": true,
    "strict": true,
    "target": "ESNext",
    "noImplicitAny": true,
    "strictNullChecks": true,
    "esModuleInterop": true
  },
  "include": ["**/*.ts", "**/*.tsx", "remix.env.d.ts", "./global.d.ts"]
}
```

**Comment by katspaugh** [comment]
Thanks! I’ll try to reproduce it with your config. In the meanwhile, a workaround would be to import it by a full path, i.e. `wavesurfer.js/dist/wavesurfer.js`.

**Comment by katspaugh** [comment]
Although actually I see it complains about full paths too…

**Comment by katspaugh** [comment]
I've just tested with beta 9 and it now resolves types correctly. Tested in a TypeScript/Next.js/Webpack project.

**Comment by katspaugh** [comment]
I've added explicit type exports, please see if beta.10 works for you. Cheers!

**Comment by jgentes** [comment]
Well the error changed :)
![image](https://github.com/katspaugh/wavesurfer.js/assets/1149470/812b9620-089c-41fb-b511-61f3e6a6354e)
```
The current file is a CommonJS module whose imports will produce 'require' calls; however, the referenced file is an ECMAScript module and cannot be imported with 'require'. Consider writing a dynamic 'import("wavesurfer.js")' call instead.
  To convert this file to an ECMAScript module, add the field `"type": "module"` to 'd:/mixpoint/package.json'.ts(1479)
```
Also it would be super helpful if the plugins exported through the main module as well, so the eventual line might read `import WaveSurfer, { MiniMapPlugin , RegionsPlugin} from 'wavesurfer.js'`

The plugins current have the original error:
![image](https://github.com/katspaugh/wavesurfer.js/assets/1149470/eb008d22-536e-403c-893d-1ec1ea7a94c1)
```
The current file is a CommonJS module whose imports will produce 'require' calls; however, the referenced file is an ECMAScript module and cannot be imported with 'require'. Consider writing a dynamic 'import("wavesurfer.js/dist/plugins/minimap.js")' call instead.
  To convert this file to an ECMAScript module, add the field `"type": "module"` to 'd:/mixpoint/package.json'.ts(1479)
```

I did verify that `node_modules` shows the right version:
![image](https://github.com/katspaugh/wavesurfer.js/assets/1149470/95983696-2cd3-48e4-8b74-3e72b930b707)

Thanks!

**Comment by jgentes** [comment]
It does look like it works fine in StackBlitz: https://stackblitz.com/edit/typescript-playground-qvt93f?file=index.ts

So I wonder if there's something odd in my config.. perhaps because I'm using Remix and I have to configure wavesurfer as a client-side dependency so that it won't load during SSR. I'll see if I can look into my config a bit more.

**Comment by jgentes** [comment]
Ok, the culprit is the `moduleResolution` option in `tsconfig.json`

If set to `"moduleResolution": "node"` the import resolves fine. If it's set to `node16` or `nodeNext` it creates the error that I am seeing.

I'm not sure what that means exactly, there's a page on it here: https://www.typescriptlang.org/tsconfig#moduleResolution that indicates the difference (CJS vs. ESM)

But I don't see any problems with using `"moduleResolution": "node"` so that appears to be a usable workaround.

**Comment by katspaugh** [comment]
I see, thanks!
This whole module business is a nightmare. :) I’m looking forward to the time when we only have ES modules.
What worries me is that NodeNext sounds like it might come back when it becomes the default. I should research and test this a bit more.

**Comment by katspaugh** [comment]
Regarding exporting modules from the the main module, I’ll think about it. We should be careful to not make people load all that code if they don’t use a tree-shaking compiler.

**Comment by benallfree** [comment]
Chiming in to report this problem with Svelte/Vite as of beta 11. 

If it helps, [@s-libs/micro-dash](https://www.npmjs.com/package/@s-libs/micro-dash) _very_ successfully implements a tree shakable import strategy. Check out their [public-api.ts](https://github.com/simontonsoftware/s-libs/blob/master/projects/micro-dash/src/public-api.ts)

**Comment by katspaugh** [comment]
@benallfree what's your tsconfig/bundler config?

**Comment by benallfree** [comment]
@katspaugh 

`tsconfig.json`:

```json
{
  "extends": "@tsconfig/svelte/tsconfig.json",
  "compilerOptions": {
    "target": "ESNext",
    "useDefineForClassFields": true,
    "module": "ESNext",
    "resolveJsonModule": true,
    /**
     * Typecheck JS in `.svelte` and `.js` files by default.
     * Disable checkJs if you'd like to use dynamic types in JS.
     * Note that setting allowJs false does not prevent the use
     * of JS in `.svelte` files.
     */
    "allowJs": true,
    "checkJs": true,
    "isolatedModules": true,
  },
  "include": ["src/**/*.d.ts", "src/**/*.ts", "src/**/*.js", "src/**/*.svelte"],
  "references": [{ "path": "./tsconfig.node.json" }]
}

```

`vite.config.ts`:

```ts
import { defineConfig } from 'vite'
import { svelte } from '@sveltejs/vite-plugin-svelte'

// https://vitejs.dev/config/
export default defineConfig({
  build: {
    minify: 'terser',
    terserOptions: {
      compress: {
        drop_console: true
      }
    }
  },
  plugins: [svelte()]
})
```

**Comment by katspaugh** [comment]
Thanks! And what does your Regions import look like?

**Comment by benallfree** [comment]
@katspaugh 

```ts
  import RegionsPlugin from 'wavesurfer.js/dist/plugins/regions.min.cjs'
```
I'm going to play with this a bit, I forked your repo.

**Comment by katspaugh** [comment]
Gotcha. That's not correct, the import should be
```
import Regions from 'wavesurfer.js/dist/plugins/regions.js'
```

**Comment by benallfree** [comment]
Ok that does work. I also have a solution in #2937 that actually builds to a smaller bundle size using Vite.

**Comment by katspaugh** [comment]
Since the types are now exported correctly, I'll go ahead and close this. Let's continue in the linked PR.

**Comment by benallfree** [comment]
Agreed. Upon reviewing everything, it seems like the existing solution is better than my proposal. 

The only mystery now is that VSCode doesn't seem to pick up the `export` section in `package.json`, so the auto import is not working.
